packages/core/src/tools/mcp-client-manager.ts

```diff
@@ -0,0 +1,245 @@
+} from './mcp-client.js';
+import { getErrorMessage } from '../utils/errors.js';
+
+/**
+ * Manages the lifecycle of multiple MCP clients, including local child processes.
+ * This class is responsible for starting, stopping, and discovering tools from
+ * a collection of MCP servers defined in the configuration.
+ */
+export class McpClientManager {
+  private clients: Map<string, Client> = new Map();
+  private localServers: Map<string, StdioClientTransport> = new Map();
+  private mcpServers: Record<string, MCPServerConfig>;
+  private mcpServerCommand: string | undefined;
+  private toolRegistry: ToolRegistry;
+  private promptRegistry: PromptRegistry;
+  private debugMode: boolean;
+  private discoveryState: MCPDiscoveryState = MCPDiscoveryState.NOT_STARTED;
+
+  constructor(
+    mcpServers: Record<string, MCPServerConfig>,
+    mcpServerCommand: string | undefined,
+    toolRegistry: ToolRegistry,
+    promptRegistry: PromptRegistry,
+    debugMode: boolean,
+  ) {
+    this.mcpServers = mcpServers;
+    this.mcpServerCommand = mcpServerCommand;
+    this.toolRegistry = toolRegistry;
+    this.promptRegistry = promptRegistry;
+    this.debugMode = debugMode;
+  }
+
+  /**
+   * Starts all configured local MCP servers as child processes.
+   * This method iterates through the server configurations, identifies those
+   * that are local (defined by a `command`), and spawns them.
+   */
+  async startLocalServers(): Promise<void> {
+    const servers = populateMcpServerCommand(
+      this.mcpServers,
+      this.mcpServerCommand,
+    );
+
+    for (const [name, config] of Object.entries(servers)) {
+      if (config.command) {
+        // This is a local server, start it as a child process
+        try {
+          const transport = new StdioClientTransport({
+            command: config.command,
+            args: config.args || [],
+            env: {
+              ...process.env,
+              ...(config.env || {}),
+            } as Record<string, string>,
+            cwd: config.cwd,
+            stderr: 'pipe',
+          });
+
+          if (this.debugMode) {
+            transport.stderr!.on('data', (data) => {
+              const stderrStr = data.toString().trim();
+              console.debug(`[DEBUG] [MCP STDERR (${name})]: `, stderrStr);
+            });
+          }
+
+          this.localServers.set(name, transport);
+        } catch (error) {
+          console.error(
+            `Failed to start local MCP server '${name}': ${getErrorMessage(
+              error,
+            )}`,
+          );
+        }
+      }
+    }
+  }
+
+  /**
+   * Initiates the tool discovery process for all configured MCP servers.
+   * It connects to each server, discovers its available tools, and registers
+   * them with the `ToolRegistry`.
+   */
+  async discoverAllMcpTools(): Promise<void> {
+    await this.startLocalServers();
+    this.discoveryState = MCPDiscoveryState.IN_PROGRESS;
+    const servers = populateMcpServerCommand(
```

Comment:
So the populate bit comes from mcp-client currently. Should it continue to come from there or be baked into this manager?